File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.py
# Last updated: 2/23/2026, 6:46:26 PM
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        def helper(node, res):
            if not node:
                return 0
            else:
                left = helper(node.left, res)
                right = helper(node.right, res)
                res[0] = max(res[0], left+right)
                return 1+max(left, right)
        res = [0]
        helper(root, res)
        return res[0]


    


        

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.java
// Last updated: 2/23/2026, 6:46:29 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int max=0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);
        return max;
    }
    
    public int maxDepth(TreeNode node) {
        if(node==null)
            return 0;
        int left = maxDepth(node.left);
        int right = maxDepth(node.right);
        max = Math.max(left+right, max);
        return 1+Math.max(left, right);
    }
}